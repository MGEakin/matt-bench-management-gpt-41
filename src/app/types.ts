import { z } from "zod";

export const EmployeeSchema = z.object({
  id: z.number().optional(),
  name: z.string(),
  title: z.string(),
  studio: z.string(),
  practice: z.string(),
  region: z.string(),
  location: z.string(),
  skill_level: z.string(),
  current_assignment: z.string().optional(),
});

export type Employee = z.infer<typeof EmployeeSchema>;
