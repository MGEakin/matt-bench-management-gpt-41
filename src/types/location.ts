import { z } from "zod";

export const LocationSchema = z.object({
  id: z.number().optional(),
  name: z.string(),
});

export type Location = z.infer<typeof LocationSchema>;
