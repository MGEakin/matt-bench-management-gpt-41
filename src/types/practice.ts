import { z } from "zod";

export const PracticeSchema = z.object({
  id: z.number(),
  name: z.string(),
});

export type Practice = z.infer<typeof PracticeSchema>;
