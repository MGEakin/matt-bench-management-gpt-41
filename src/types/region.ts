import { z } from "zod";

export const RegionSchema = z.object({
  id: z.number().optional(),
  name: z.string(),
});

export type Region = z.infer<typeof RegionSchema>;
