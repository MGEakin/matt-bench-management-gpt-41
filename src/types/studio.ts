import { z } from "zod";

export const StudioSchema = z.object({
  id: z.number().optional(),
  name: z.string(),
});

export type Studio = z.infer<typeof StudioSchema>;
