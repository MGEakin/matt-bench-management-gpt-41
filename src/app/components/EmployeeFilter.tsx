import React, { useEffect, useState } from "react";
import { Practice } from "../../types/practice";
import { Studio } from "../../types/studio";
import { fetchPractices } from "../../lib/fetchPractices";
import { fetchStudios } from "../../lib/fetchStudios";

interface EmployeeFilterProps {
  filters: Record<string, string>;
  onChange: (filters: Record<string, string>) => void;
}

const fields = [
  "name",
  "title",
  "studio",
  "practice",
  "region",
  "location",
  "skill_level",
  "current_assignment",
];

const EmployeeFilter: React.FC<EmployeeFilterProps> = ({ filters, onChange }) => {
  const [practices, setPractices] = useState<Practice[]>([]);
  const [loadingPractices, setLoadingPractices] = useState(true);
  const [errorPractices, setErrorPractices] = useState<string | null>(null);

  const [studios, setStudios] = useState<Studio[]>([]);
  const [loadingStudios, setLoadingStudios] = useState(true);
  const [errorStudios, setErrorStudios] = useState<string | null>(null);

  useEffect(() => {
    let mounted = true;
    setLoadingPractices(true);
    fetchPractices()
      .then((data) => {
        if (mounted) {
          setPractices(data);
          setLoadingPractices(false);
        }
      })
      .catch((err) => {
        if (mounted) {
          setErrorPractices(err.message);
          setLoadingPractices(false);
        }
      });
    return () => {
      mounted = false;
    };
  }, []);

  useEffect(() => {
    let mounted = true;
    setLoadingStudios(true);
    fetchStudios()
      .then((data) => {
        if (mounted) {
          setStudios(data);
          setLoadingStudios(false);
        }
      })
      .catch((err) => {
        if (mounted) {
          setErrorStudios(err.message);
          setLoadingStudios(false);
        }
      });
    return () => {
      mounted = false;
    };
  }, []);

  return (
    <form
      className="flex flex-wrap gap-4 mb-6"
      role="form"
      onSubmit={(e) => e.preventDefault()}
    >
      {/* Render all fields except practice and studio as text inputs */}
      {fields.filter((field) => field !== "practice" && field !== "studio").map((field) => (
        <input
          key={field}
          type="text"
          placeholder={`Filter by ${field.replace("_", " ")}`}
          value={filters[field] || ""}
          onChange={(e) => onChange({ ...filters, [field]: e.target.value })}
          className="border rounded px-2 py-1 dark:bg-gray-800 dark:text-white"
        />
      ))}
      {/* Studio dropdown */}
      <div>
        {loadingStudios ? (
          <select className="border rounded px-2 py-1 dark:bg-gray-800 dark:text-white" disabled>
            <option>Loading studios...</option>
          </select>
        ) : errorStudios ? (
          <select className="border rounded px-2 py-1 dark:bg-gray-800 dark:text-white" disabled>
            <option>Error loading studios</option>
          </select>
        ) : (
          <select
            className="border rounded px-2 py-1 dark:bg-gray-800 dark:text-white"
            value={filters.studio || ""}
            onChange={(e) => onChange({ ...filters, studio: e.target.value })}
          >
            <option value="">Filter by studio</option>
            {studios.map((studio) => (
              <option key={studio.id} value={studio.name}>
                {studio.name}
              </option>
            ))}
          </select>
        )}
      </div>
      {/* Practice dropdown */}
      <div>
        {loadingPractices ? (
          <select className="border rounded px-2 py-1 dark:bg-gray-800 dark:text-white" disabled>
            <option>Loading practices...</option>
          </select>
        ) : errorPractices ? (
          <select className="border rounded px-2 py-1 dark:bg-gray-800 dark:text-white" disabled>
            <option>Error loading practices</option>
          </select>
        ) : (
          <select
            className="border rounded px-2 py-1 dark:bg-gray-800 dark:text-white"
            value={filters.practice || ""}
            onChange={(e) => onChange({ ...filters, practice: e.target.value })}
          >
            <option value="">Filter by practice</option>
            {practices.map((practice) => (
              <option key={practice.id} value={practice.name}>
                {practice.name}
              </option>
            ))}
          </select>
        )}
      </div>
    </form>
  );
};

export default EmployeeFilter;
