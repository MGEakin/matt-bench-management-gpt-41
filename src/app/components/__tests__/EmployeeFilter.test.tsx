import React from "react";
import { describe, it, expect, vi } from "vitest";
import { render, screen, fireEvent, waitFor } from "@testing-library/react";
import "@testing-library/jest-dom";
import { Practice } from "../../../types/practice";
import { Studio } from "../../../types/studio";
const practices: Practice[] = [
  { id: 1, name: "Engineering" },
  { id: 2, name: "Design" },
  { id: 3, name: "Strategy" },
];
const studios: Studio[] = [
  { id: 1, name: "NY" },
  { id: 2, name: "SF" },
  { id: 3, name: "London" },
];
vi.stubGlobal('fetch', (url: string) => {
  if (url.includes('/api/practices')) {
    return Promise.resolve({
      ok: true,
      json: () => Promise.resolve(practices),
    });
  }
  if (url.includes('/api/studios')) {
    return Promise.resolve({
      ok: true,
      json: () => Promise.resolve(studios),
    });
  }
  return Promise.reject(new Error('Unknown endpoint'));
});
import EmployeeFilter from "../EmployeeFilter";

import { create } from "react-test-renderer";

describe("EmployeeFilter", () => {
  it("renders filter inputs, studio dropdown, and practice dropdown", async () => {
    render(<EmployeeFilter filters={{}} onChange={() => {}} />);
    expect(screen.getByPlaceholderText("Filter by name")).toBeInTheDocument();
    expect(screen.getByPlaceholderText("Filter by title")).toBeInTheDocument();
    // Wait for studios and practices to load
    await waitFor(() => {
      const selects = screen.getAllByRole("combobox");
      expect(selects.length).toBe(2);
      expect(selects[0]).toHaveDisplayValue("Filter by studio");
      expect(selects[1]).toHaveDisplayValue("Filter by practice");
    });
    for (const studio of studios) {
      expect(screen.getByRole("option", { name: studio.name })).toBeInTheDocument();
    }
    for (const practice of practices) {
      expect(screen.getByRole("option", { name: practice.name })).toBeInTheDocument();
    }
  });

  it("calls onChange when input values change including studio dropdown", async () => {
    const onChange = vi.fn();
    render(<EmployeeFilter filters={{}} onChange={onChange} />);
    const nameInput = screen.getByPlaceholderText("Filter by name");
    const titleInput = screen.getByPlaceholderText("Filter by title");
    // Wait for studios and practices to load
    let selects: HTMLElement[] = [];
    await waitFor(() => {
      selects = screen.getAllByRole("combobox");
      expect(selects.length).toBe(2);
    });
    fireEvent.change(nameInput, { target: { value: "John" } });
    fireEvent.change(titleInput, { target: { value: "Manager" } });
    fireEvent.change(selects[0], { target: { value: "NY" } });
    fireEvent.change(selects[1], { target: { value: "Engineering" } });
    expect(onChange).toHaveBeenCalled();
    expect(onChange).toHaveBeenCalledTimes(4);
  });

  it("renders with all filters applied including studio and practice", async () => {
    render(
      <EmployeeFilter
        filters={{ name: "Alice", title: "QA", studio: "NY", practice: "Design" }}
        onChange={() => {}}
      />,
    );
    expect(screen.getByDisplayValue("Alice")).toBeInTheDocument();
    expect(screen.getByDisplayValue("QA")).toBeInTheDocument();
    // Wait for studios and practices to load
    let selects: HTMLElement[] = [];
    await waitFor(() => {
      selects = screen.getAllByRole("combobox");
      expect(selects.length).toBe(2);
      expect(selects[0]).toHaveValue("NY");
      expect(selects[1]).toHaveValue("Design");
    });
  });

  it("matches snapshot", async () => {
    const tree = create(
      <EmployeeFilter filters={{}} onChange={() => {}} />,
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("has accessible form", async () => {
    render(<EmployeeFilter filters={{}} onChange={() => {}} />);
    expect(screen.getByRole("form")).toBeInTheDocument();
    const selects = screen.getAllByRole("combobox");
    expect(selects.length).toBe(2);
    selects.forEach(select => expect(select).toBeInTheDocument());
  });
});
