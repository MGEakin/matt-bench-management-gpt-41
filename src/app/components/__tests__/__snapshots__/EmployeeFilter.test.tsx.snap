// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`EmployeeFilter > matches snapshot 1`] = `null`;
